controller/src/main/java/io/pravega/controller/server/bucket/BucketManager.java

```diff
@@ -54,37 +54,83 @@ public abstract class BucketManager extends AbstractService {
     private final Map<Integer, BucketService> buckets;
     @Getter(AccessLevel.PROTECTED)
     private final ScheduledExecutorService executor;
-
+    private final BucketStore bucketStore;
+    /**
+     * The tests can add listeners to get notification when the update has happened in the store.
+     */
+    @Getter
+    private final AtomicReference<BucketListener> listenerRef;
     BucketManager(final String processId, final BucketStore.ServiceType serviceType, final ScheduledExecutorService executor,
-                  final Function<Integer, BucketService> bucketServiceSupplier) {
+                  final Function<Integer, BucketService> bucketServiceSupplier, final BucketStore bucketStore) {
         this.processId = processId;
         this.serviceType = serviceType;
         this.executor = executor;
         this.lock = new Object();
         this.buckets = new HashMap<>();
         this.bucketServiceSupplier = bucketServiceSupplier;
+        this.bucketStore = bucketStore;
+        listenerRef = new AtomicReference<>();
     }
 
     @Override
     protected void doStart() {
-        initializeService()
-                .thenCompose(s ->
-                        Futures.allOf(IntStream.range(0, getBucketCount()).boxed().map(x -> initializeBucket(x)
-                                .thenCompose(v -> tryTakeOwnership(x))).collect(Collectors.toList())))
-                .thenAccept(x -> startBucketOwnershipListener())
-                .whenComplete((r, e) -> {
-                    if (e != null) {
-                        notifyFailed(e);
-                    } else {
-                        notifyStarted();
-                    }
-                });
+        initializeService().thenAccept(v -> addBucketControllerMapListener())
+                           .thenAccept(v -> startLeaderElection())
+                           .thenCompose(v -> bucketStore.getBucketsForController(processId, serviceType))
+                           .thenCompose(buckets -> Futures.allOf(buckets.stream().map(x -> initializeBucket(x)
+                                                                                .thenCompose(v -> tryTakeOwnership(x)))
+                                                                        .collect(Collectors.toList())))
+                           .thenAccept(x -> startBucketOwnershipListener())
+                           .whenComplete((r, e) -> {
+                               if (e != null) {
+                                   notifyFailed(e);
+                               } else {
+                                   notifyStarted();
+                               }
+                           });
     }
 
     protected abstract int getBucketCount();
 
     public abstract boolean isHealthy();
 
+    public abstract void startLeaderElection();
+
+    public abstract void startLeader();
+
+    public abstract void stopLeader();
+
+    protected CompletableFuture<Void> manageBuckets(int noOfClusters) {
```

Comment:
Changed ` noOfClusters` -> `numberOfClusters` and also changed Cluster to Controller. Added java doc for this method. This method invokes whenever there is any change in existing bucket controller mapping. Depending on situation either it will release buckets(if new controller instance is added) or acquire buckets(if any existing controller instance failed). Depending on this i name it as `manageBuckets`.